python/ray/data/datasource/file_meta_provider.py

```diff
@@ -365,3 +364,162 @@ def _handle_read_os_error(error: OSError, paths: Union[str, List[str]]) -> str:
         )
     else:
         raise error
+
+
+def _expand_paths(
+    paths: List[str],
+    filesystem: "pyarrow.fs.FileSystem",
+    partitioning: Optional[Partitioning],
+) -> Iterator[Tuple[str, int]]:
+    """Get the file sizes for all provided file paths."""
+    from pyarrow.fs import LocalFileSystem
+    from ray.data.datasource.file_based_datasource import (
+        _wrap_s3_serialization_workaround,
+        _unwrap_protocol,
+    )
+
+    # We break down our processing paths into a few key cases:
+    # 1. If len(paths) < threshold, fetch the file info for the individual files/paths
+    #    serially.
+    # 2. If all paths are contained under the same parent directory (or base directory,
+    #    if using partitioning), fetch all file infos at this prefix and filter to the
+    #    provided paths on the client; this should be a single file info request.
+    # 3. If more than threshold requests required, parallelize them via Ray tasks.
+
+    # 1. Small # of paths case.
+    if (
+        len(paths) < FILE_SIZE_FETCH_PARALLELIZATION_THRESHOLD
+        # Local file systems are very fast to hit.
+        or isinstance(filesystem, LocalFileSystem)
+    ):
+        for path in paths:
+            yield from _get_file_infos(path, filesystem)
+        return
+
+    # 2. Common path prefix case.
+    # Get longest common path of all paths.
+    common_path = os.path.commonpath(paths)
+    common_path = os.path.normpath(common_path)
+    # If parent directory (or base directory, if using partitioning) is common to all
+    # paths, fetch all file infos at that prefix and filter the response to the provided
+    # paths.
+    normalized_base_dir = _unwrap_protocol(partitioning.base_dir)
+    if (partitioning is not None and common_path == normalized_base_dir) or all(
+        str(pathlib.Path(path).parent) == common_path for path in paths
+    ):
+        path_to_size = {path: 0 for path in paths}
+        for path, file_size in _get_file_infos(common_path, filesystem):
+            if path in path_to_size:
+                path_to_size[path] = file_size
+        # Dictionaries are insertion-ordered, so this path + size pairs should be
+        # yielded in the order of the original paths arg.
+        yield from path_to_size.items()
+        return
+
+    # 3. Parallelization case.
+    # Parallelize requests via Ray tasks.
+    # TODO(Clark): Group file paths by parent directory paths and do a prefix
+    # fetch + client-side filter if the number of groups is << the total number of
+    # paths?
+    # Only request 0.25 CPU since these tasks are I/O-bound.
+    num_cpus = 0.25
```

Comment:
Can we pull out the metadata parallelization / remote task routine into a common one between the parquet datasource and file-based one? There are small differences like this that make it hard to maintain.

Btw, we should probably use at fewest 0.5 CPUs per task, to avoid the nasty too many workers warnings.

Reply:
Hmm I was thinking that deviation was desirable here since (1) fetching Parquet metadata (which involves actually reading a metadata file) is faster than fetching the file size so we'd want the number of fetches-per-task to differ between the two, and (2) the actual fetching tasks are different and involve difference identifiers (Parquet fragments vs. file paths). I was also hoping to not change the Parquet metadata fetching behavior in this PR, if possible.

I'll take a stab at consolidating these as much as possible.

>Btw, we should probably use at fewest 0.5 CPUs per task, to avoid the nasty too many workers warnings.

I got far worse (2x) performance with 0.5 CPUs per task on a few different benchmark configs (128 files, 1024 files, 10k files, 40k files), and if I were a user I'd trade off a few of those warnings for that large of a perf improvement.

But I'll change this to 0.5 CPUs for now, since I really don't want to delay getting this merged on this.